dbms/src/Flash/Mpp/FineGrainedShuffleWriter.cpp

```diff
@@ -57,81 +58,105 @@ void FineGrainedShuffleWriter<StreamWriterPtr>::finishWrite()
+    if (cached_block_count == fine_grained_shuffle_stream_count || static_cast<UInt64>(rows_in_blocks) >= fine_grained_shuffle_batch_size * fine_grained_shuffle_stream_count)
         batchWriteFineGrainedShuffle<false>();
 }
 
+template <class StreamWriterPtr>
+void FineGrainedShuffleWriter<StreamWriterPtr>::resetScatterColumns()
+{
+    scattered.resize(num_columns);
+    for (size_t col_id = 0; col_id < num_columns; ++col_id)
+    {
+        auto & column = header.getByPosition(col_id).column;
+
+        scattered[col_id].reserve(num_bucket);
+        for (size_t chunk_id = 0; chunk_id < num_bucket; ++chunk_id)
+        {
+            scattered[col_id].emplace_back(column->cloneEmpty());
+            scattered[col_id][chunk_id]->reserve(1024);
+        }
+    }
+}
+
 template <class StreamWriterPtr>
 template <bool send_exec_summary_at_last>
 void FineGrainedShuffleWriter<StreamWriterPtr>::batchWriteFineGrainedShuffle()
 {
     std::vector<TrackedMppDataPacket> tracked_packets(partition_num);
-
     if (!blocks.empty())
     {
         assert(rows_in_blocks > 0);
         assert(fine_grained_shuffle_stream_count <= 1024);
 
-        // fine_grained_shuffle_stream_count is in (0, 1024], and partition_num is uint16_t, so will not overflow.
-        uint32_t bucket_num = partition_num * fine_grained_shuffle_stream_count;
-
         HashBaseWriterHelper::materializeBlocks(blocks);
-        auto final_dest_tbl_columns = HashBaseWriterHelper::createDestColumns(blocks[0], bucket_num);
-        Block dest_block = blocks[0].cloneEmpty();
-
-        // Hash partition input_blocks into bucket_num.
-        while (!blocks.empty())
+        if (!inited)
         {
-            const auto & block = blocks.back();
-            size_t columns = block.columns();
-            std::vector<String> partition_key_containers(collators.size());
-            auto dest_tbl_columns = HashBaseWriterHelper::createDestColumns(block, bucket_num);
-            HashBaseWriterHelper::computeHash(block, bucket_num, collators, partition_key_containers, partition_col_ids, dest_tbl_columns);
-            blocks.pop_back();
-
-            for (size_t bucket_idx = 0; bucket_idx < bucket_num; ++bucket_idx)
-            {
-                for (size_t col_id = 0; col_id < columns; ++col_id)
-                {
-                    const MutableColumnPtr & src_col = dest_tbl_columns[bucket_idx][col_id];
-                    final_dest_tbl_columns[bucket_idx][col_id]->insertRangeFrom(*src_col, 0, src_col->size());
-                }
-            }
+            header = blocks[0].cloneEmpty();
+            num_columns = header.columns();
+            // fine_grained_shuffle_stream_count is in (0, 1024], and partition_num is uint16_t, so will not overflow.
+            num_bucket = partition_num * fine_grained_shuffle_stream_count;;
+            partition_key_containers_for_reuse.resize(collators.size());
+            resetScatterColumns();
+            inited = true;
```

Comment:
Can we move to `init(const Block & header)` and Called by `readPrefixImpl`?

Reply:
Nice suggestion, I'll try.

Reply:
Done.